### DISCUSSION TITLE
2.66 inch Pico ePaper Module with nano-gui

### DISCUSSION THREAD

**Discussion by lkoehl** [micropython/micropython#11479] [Hardware & Peripherals] [answered]
I am trying to get a 2.66 inch ePaper module from waveshare to work with custom fonts and a Raspberry Pico. It came with a sample project that I found [here](https://github.com/waveshare/Pico_ePaper_Code/blob/main/python/Pico_ePaper-2.66.py). Unfortunately, as I found out, the framebuffer in MicroPython does not support different font sizes. Looking further, I found the fantastic [nano gui project](https://github.com/peterhinch/micropython-nano-gui) and tinkered with it. The project only provides drivers for the 2.9 inch epaper screen. Thinking that the differences between the 2.9 and 2.66 screens could not be that big, I changed some smaller things like the height. Unfortunately, the screen remained empty. 
As the sample project for the 2.9 inch screen from waveshare works on the 2.66 inch screen, with the obvious drawback that you could only draw in the 'cropped' area. As this was the case, I tried the standard drivers for the 2.9 inch screen and the screen was still blank.
I also noticed that waveshare offers a [sample project with nano gui](https://github.com/waveshare/Pico_ePaper_Code/blob/main/pythonNanoGui/drivers/ePaper2in9.py), unfortunately only for the 2.9 inch screen. Using this I was able to draw some pixel garbage on the screen, but unfortunately it does not reassamble what should be drawn.
 
As the low level commands are a bit beyond my level, do you guys have any ideas on how to modify the driver for the 2.66 inch screen or any ideas on what I am overlooking?

**Comment by peterhinch** [comment]
Supporting a new display is non trivial. First I have to buy a unit, then study the Waveshare code and port it to MicroPython. I know some Waveshare drivers are already in MP, but their code is such that producing something worthy of release constitutes days of work.

By all means experiment, but my efforts are on [this display](https://www.waveshare.com/pico-epaper-4.2.htm) which provides a number of great facilities. Notably partial updates (supported) and greyscale (WIP). I recommend new users of ePaper to use that display.

**Reply by lkoehl** [reply]
I spent some more time trying to modify the 2.9 inch driver to my needs by looking at all the available commands in the [specification](https://www.waveshare.com/w/upload/d/dc/2.66inch-e-paper-specification.pdf). 

I tried to document and understand all the commands in the driver. Eventually I got the seismograph example to work, with one small drawback: everything is mirrored.

![IMG_5976](https://github.com/micropython/micropython/assets/48966896/0c98e599-e87e-483e-9e5b-b0035f86224d)

Since everything is drawn on the screen, I suspect there is a wrong value somewhere. 

[This](https://gist.github.com/lkoehl/f378807104955d61a99f9157e14423ec) is the modified driver that produces the above results. Maybe you have an idea of what the problem might be.

**Comment by peterhinch** [comment]
I will find time to look at this, but it may b a couple of days before I can do it.

There are two problems with unsupported displays. One is where they use a different controller chip. I suspect this isn't the case given the results you have achieved. But the other problem is where the same controller chip is used, but is wired to the display hardware in a different way. I suspect that may be why it is mirrored.

In general to fix a mirror-image display you have to reverse the order in which rows are output, which reflects the image about the horizontal axis, and, in each row, reverse the order in which pixels are output. This reflects the image about a vertical axis. You may be able to achieve this with the data entry sequence command in the spec.

**Reply by lkoehl** [reply]
Thank you very much for having a look. 
I experimented with the data entry sequence command, but had no success.

**Reply by peterhinch** [reply]
The other approach is to do it in Python adapting the display driver. This can be more tricky than tweaking the chip by changing the data entry sequence.

**Comment by peterhinch** [comment, marked as answer]
I've now studied your driver and the [waveshare driver](https://github.com/waveshare/Pico_ePaper_Code/blob/f6af2a819d1181a1629321a3ff3aaaf0b82e0fe0/python/Pico_ePaper-2.66.py#L213). You'll note that the Waveshare driver outputs pixels in right-to-left order, which may well explain the mirror image about the vertical axis.

If I were writing a driver for this display I would start from the Waveshare code rather than from my driver for the 2.9" Adafruit display, which uses a completely different controller chip (the IL0373). I think copying Waveshare's right-to-left rendering should fix your problem.

I have taken this opportunity to update the 2.9" driver. It will now run demos intended for non-EPD displays (albeit with a lot of display refreshes).

**Reply by lkoehl** [reply]
Thanks for helping out. Reversing the output order results in a correctly drawn image.

```py
if self._lsc:  # Landscape mode
    wid = self.width
    tbc = self.height // 8  # Vertical bytes per column
    iidx = wid * (tbc - 1)  # Initial index
    idx = iidx  # Index into framebuf
    vbc = 0  # Current vertical byte count
    hpc = self.width - 1 # Horizontal pixel count
    
    for _ in range(len(mvb)):
        self._cs(0)
        buf1[0] = ~mvb[idx]  # INVERSION HACK ~data
        send(buf1)
        self._cs(1)
        idx -= self.width
        vbc += 1
        vbc %= tbc
        if not vbc:
            hpc -= 1
            idx = iidx + hpc  
```
 It probably needs to be changed in `async def _as_show(self, buf1=bytearray(1)):` aswell

**Reply by peterhinch** [reply]
I'm glad you've got it working. The 2.9" driver is rather hard to follow because the unit's natural orientation is portrait. The rotation is done in code. Evidently the Waveshare unit does landscape natively.

Does your driver support portrait mode? If you want to post a complete driver I'd be happy to include it (crediting you as author, of course).

**Reply by lkoehl** [reply]
I just tested out portrait mode by taking a look at the [waveshare driver.](https://github.com/waveshare/Pico_ePaper_Code/blob/f6af2a819d1181a1629321a3ff3aaaf0b82e0fe0/python/Pico_ePaper-2.66.py#LL191C5-L199C29)
There it is implemented by just iterating over the image buffer.
```py
def display(self, image):
    if (image == None):
        return            
        
    self.SetCursor(1, 295)
    
    self.send_command(0x24) # WRITE_RAM
    self.send_data1(image)
    self.TurnOnDisplay()
```

Trying the same in my driver, results again in a vertical flipped image.
```py
for b in mvb:
    self._cs(0)
    buf1[0] = ~b  # INVERSION HACK ~data
    send(buf1)
    self._cs(1)
```

Its possible to flip the image in code again, but it makes my think, that the driver chip isn't configured properly.

**Reply by lkoehl** [reply]
You can find the complete code [here](https://gist.github.com/lkoehl/f378807104955d61a99f9157e14423ec#file-epaper2in66-py).
I don't have any experience with async operation and the portrait image is currently flipped.

**Reply by peterhinch** [reply]
OK, thanks for the code. I'll figure out whether to buy a unit and complete your driver's development. I'm currently trying to complete a greyscale driver for the 4.2" display. The 2-color ones look interesting, too...

**Reply by fullmanido** [reply]
Thanks so much for the help! I'm using the 2.66 Pico HAT without much success. Are you using fonts or an image? Would it be possible to upload the code, we are doing a scout project with solar. Many thanks!

**Reply by peterhinch** [reply]
@fullmanido I don't plan to develop the 2.66" driver because the 4.2" display is much better, supporting 4-level greyscale or partial (i.e. fast) updates. 

The driver posted by @lkoehl works with [nano-gui](https://github.com/peterhinch/micropython-nano-gui) in landscape mode. If you only want to display text you can use the [Writer class](https://github.com/peterhinch/micropython-font-to-py/blob/master/writer/WRITER.md) - this is a component of nano-gui, so both allow text display in your choice of font and size.

**Reply by lkoehl** [reply]
@peterhinch I also bought the 4.2" display as you recommended. While it might be better, it absolutely didn't fit my use case because of its size. The 2.66" screen has a similar size as the 2" screen, while having a higher resolution.

**Reply by peterhinch** [reply]
If you wish I'm happy to add your driver to the `nano-gui` drivers (crediting you with its development). I agree there is a use case for it, and your driver evidently works fine. 

The point I was trying to make is that further development is possible (portrait mode, greyscales) - but I have no plans to take this on.

Which display did you test on? [The Pico version](https://www.waveshare.com/pico-epaper-2.66.htm) or [the universal one](https://www.waveshare.com/product/displays/e-paper/epaper-2/2.66inch-e-paper-module.htm)?
